vxquery-core/src/main/java/org/apache/vxquery/hdfs2/HDFSFunctions.java

```diff
@@ -0,0 +1,463 @@
+        HashMap<String, ArrayList<Integer>> splits_map = this.getLocationsOfSplits();
+        readNodesFromXML();
+        int count = this.splits.size();
+
+        ArrayList<Integer> splits;
+        String node;
+        for (ArrayList<String> info : this.nodes) {
+            node = info.get(0);
+            if (splits_map.containsKey(node)) {
+                splits = splits_map.get(node);
+                for (Integer split : splits) {
+                    schedule.put(split, node);
+                    count--;
+                }
+                splits_map.remove(node);
+            } else {
+                empty.add(node);
+            }
+        }
+
+        //Check if every split got assigned to a node
+        if (count != 0) {
+            ArrayList<Integer> remaining = new ArrayList<Integer>();
+            // Find remaining splits
+            for (InputSplit s : this.splits) {
+                int i = 0;
+                if (!schedule.containsKey(i)) {
+                    remaining.add(i);
+                }
+            }
+
+            if (empty.size() != 0) {
+                int node_number = 0;
+                for (int split : remaining) {
+                    if (node_number == empty.size()) {
+                        node_number = 0;
+                    }
+                    schedule.put(split, empty.get(node_number));
+                    node_number++;
+                }
+            }
+        }
+        // TODO remove from here this is for debugging only
+        for (int s : schedule.keySet()) {
+            System.out.println("split: " + s + ", host: " + schedule.get(s));
+        }
+    }
+
+    /**
+     * Read the hostname and the ip address of every node from the xml cluster configuration file.
+     * Save the information inside an ArrayList.
+     */
+    public void readNodesFromXML() {
+        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder dBuilder;
+        try {
+            dBuilder = dbFactory.newDocumentBuilder();
+            Document doc = dBuilder.parse(nodeXMLfile);
+            doc.getDocumentElement().normalize();
+
+            nodes = new ArrayList<ArrayList<String>>();
+            NodeList nList = doc.getElementsByTagName("node");
+
+            for (int temp = 0; temp < nList.getLength(); temp++) {
+
+                Node nNode = nList.item(temp);
+
+                if (nNode.getNodeType() == Node.ELEMENT_NODE) {
+
+                    Element eElement = (Element) nNode;
+                    ArrayList<String> info = new ArrayList<String>();
+                    info.add(eElement.getElementsByTagName("id").item(0).getTextContent());
+                    info.add(eElement.getElementsByTagName("cluster_ip").item(0).getTextContent());
+                    nodes.add(info);
+                }
+            }
+        } catch (ParserConfigurationException e) {
+            System.err.println(e);
```

Comment:
The whole file prints the error messages. Is that the best practice? Should these go to a logger? Or let the exception through?